async-http-client/src/main/scala/org/http4s/client/asynchttpclient/AsyncHttpClient.scala

```diff
@@ -0,0 +1,134 @@
+package org.http4s
+package client
+package asynchttpclient
+
+import org.asynchttpclient.AsyncHandler.State
+import org.asynchttpclient.request.body.generator.{InputStreamBodyGenerator, BodyGenerator}
+import org.asynchttpclient.{Request => AsyncRequest, Response => _, _}
+import org.asynchttpclient.handler.StreamedAsyncHandler
+
+import org.http4s.util.task
+import org.reactivestreams.{Subscription, Subscriber, Publisher}
+import scodec.bits.ByteVector
+
+import scala.collection.JavaConverters._
+
+import scalaz.stream.io._
+import scalaz.stream.async
+import scalaz.concurrent.Task
+
+import scala.concurrent.Promise
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import org.log4s.getLogger
+
+object AsyncHttpClient {
+  private[this] val log = getLogger
+
+  val defaultConfig = new DefaultAsyncHttpClientConfig.Builder()
+    .setMaxConnectionsPerHost(200)
+    .setMaxConnections(400)
+    .setRequestTimeout(30000)
+    .build()
+
+  def apply(config: AsyncHttpClientConfig = defaultConfig): Client = {
+    val client = new DefaultAsyncHttpClient(config)
+    Client(Service.lift { req =>
+      val p = Promise[DisposableResponse]
+      client.executeRequest(toAsyncRequest(req), asyncHandler(p))
+      task.futureToTask(p.future)
+    }, Task(client.close()))
+  }
+
+  private def asyncHandler(promise: Promise[DisposableResponse]): AsyncHandler[Unit] =
+    new StreamedAsyncHandler[Unit] {
+      var state: State = State.CONTINUE
+      val queue = async.boundedQueue[ByteVector](1)
+
+      var disposableResponse = DisposableResponse(Response(body = queue.dequeue), Task {
+        state = State.ABORT
+        queue.close.run
+      })
+
+      override def onStream(publisher: Publisher[HttpResponseBodyPart]): State = {
+        publisher.subscribe(new Subscriber[HttpResponseBodyPart] {
+          var subscription: Option[Subscription] = None
+
+          override def onError(t: Throwable): Unit = {
+            subscription = None
+            queue.fail(t).run
+          }
+
+          override def onSubscribe(s: Subscription): Unit = {
+            subscription = Some(s)
+            s.request(1)
+          }
+
+          override def onComplete(): Unit = {
+            subscription = None
+            queue.close.run
+          }
+
+          override def onNext(t: HttpResponseBodyPart): Unit = {
+            subscription foreach { s =>
+              state match {
+                case State.CONTINUE =>
+                  queue.enqueueOne(ByteVector(t.getBodyPartBytes)).run
```

Comment:
I don't think it will block.  `onNext` can't be called more than the number of times it is requested on the subscription.  If more requests aren't available, it should set the readable flag on the channel and apply backpressure by TCP.

We might consider requesting more than 1 chunk at a time.  That may be more efficient, but trades away space.  How much space?  Who knows, we don't know the chunk size! :fearful: 

If my debugging was right, the subscriber on the request body requests 16 chunks at a time.  Maybe this is an appropriate Ignorant Default.